bin/varnishd/cache/cache_backend.c

```diff
@@ -233,8 +234,10 @@ vbe_dir_getfd(VRT_CTX, struct worker *wrk, VCL_BACKEND dir, struct backend *bp,
 			VTAILQ_REMOVE(&bp->cw_head, cw, cw_list);
```

Comment:
Not directly related to this PR, but related to the problem we are trying to solve: In the line above, is `err != 0` actually correct?
IIUC, if we got signaled (`err == 0`), we could race for the `mtx` and find the backend already busy again. But we only look at the busy state if waiting for the condition variable resulted in an error (likely `ETIMEDOUT`), so we would go ahead trying to make a connection even if the backend is busy (again).

Reply:
FYI, this has changed a bit: https://github.com/varnishcache/varnish-cache/commit/e46f97278f1036f4e648dd64d3f34ada8d29f64a (I am planning to rebase soon)
But, your statement still holds true. Maybe this should have been an OR?

The reason for this change came from you starting this thread: https://github.com/varnishcache/varnish-cache/pull/4030#discussion_r1494784649

Reply:
should this rather be something like `do { ... } while (VRT_Healthy && BE_BUSY(...) && VTIM_real() < wait_end`?

Reply:
Yes, something like this.
This is normally how we re-check the condition after condwait, and I think that is cleaner as well (easy to understand what happens).
We would need to recalculate the timeout, though.

Reply:
Nit: I would only put one condition in the while loop and the rest as if statement with breaks in the loop.